nighthawk/source/common/sequencer_impl.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include "common/common/logger.h"
+
+#include "envoy/common/pure.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/event/timer.h"
+#include "envoy/runtime/runtime.h"
+
+#include "nighthawk/common/rate_limiter.h"
+#include "nighthawk/common/sequencer.h"
+
+#include "nighthawk/source/common/statistic_impl.h"
+
+namespace Nighthawk {
+
+class SequencerImpl : public Sequencer, public Envoy::Logger::Loggable<Envoy::Logger::Id::main> {
+public:
+  SequencerImpl(Envoy::Event::Dispatcher& dispatcher, Envoy::TimeSource& time_source,
+                RateLimiter& rate_limiter, SequencerTarget& target,
+                std::chrono::microseconds duration, std::chrono::microseconds grace_timeout);
+  void start() override;
+  void waitForCompletion() override;
+
+  // TODO(oschaaf): calling this after stop() will return broken/unexpected results.
+  double completions_per_second() override {
+    double us =
+        std::chrono::duration_cast<std::chrono::microseconds>(time_source_.monotonicTime() - start_)
+            .count();
+
+    return us == 0 ? 0 : ((targets_completed_ / us) * 1000000);
+  }
+
+  // TODO(oschaaf): we want to track time we wait between the rate limiter
+  // indicating we should call target_() but target returs false, meaning
+  // we are blocked / have entered closed loop mode.
+  const HdrStatistic& blocked_statistic() override { return blocked_statistic_; }
+  const HdrStatistic& latency_statistic() override { return latency_statistic_; }
+
+  // Spinning makes tests hang when simulated time is used.
```

Comment:
Can't we avoid this by mocking out the `pthread_yield` using the `Api` layer in Envoy? This would allow us to increment simulated time.